File: module_6_c++_voor_industriele_toepassingen/les_5_templates/prog_03_matrix/matrix.cpp

template <class T>
Matrix <T> ::Matrix (int nrOfRows, int nrOfColumns):
    nrOfRows (nrOfRows),
    nrOfColumns (nrOfColumns),
    entries (nrOfRows * nrOfColumns)
{}
    
template <class T>
Matrix <T> ::Matrix (int nrOfRows, st::vector <T> const &entries):
    nrOfRows (nrOfRows),
    nrOfColumns (entries.size () / nrOfRows),
    entries (entries)
{}

template <class T>
Matrix <T> ::Matrix (Matrix <T> const &matrix):
    nrOfRows (matrix.nrOfRows),
    nrOfColumns (matrix.nrOfColumns),
    entries (matrix.entries)
{}

template <class T>
T &Matrix <T>::access (int rowIndex, int columnIndex) {
    return entries [rowIndex * nrOfColumns + columnIndex];
}

template <class T>
void Matrix <T> ::print (st::ostream &outputStream) {
    outputStream << '\n';
    for (auto rowIndex = 0; rowIndex < nrOfRows; rowIndex++) {
        for (auto columnIndex = 0; columnIndex < nrOfColumns; columnIndex++) {
            outputStream << st::setw (10) << this->access (rowIndex, columnIndex);
        }
        outputStream << '\n';
    }
    outputStream << '\n';
}

template <class U>
Matrix <U> add (Matrix <U> &matrix0, Matrix <U> &matrix1) {
    auto result = Matrix <U> (matrix0);
    for (auto entryIndex = 0; entryIndex < result.entries.size (); entryIndex++) {
        result.entries [entryIndex] += matrix1.entries [entryIndex];
    }
    return result;
}

template <class U>
Matrix <U> multiply (Matrix <U> &matrix0, Matrix <U> &matrix1) {
    auto result = Matrix <U> (matrix0.nrOfRows, matrix1.nrOfColumns);
    for (auto rowIndex = 0; rowIndex < result.nrOfRows; rowIndex++) {
        for (auto columnIndex = 0; columnIndex < result.nrOfColumns; columnIndex++) {
            for (auto termIndex = 0; termIndex < matrix0.nrOfColumns; termIndex++) {
                result.access (rowIndex, columnIndex) += matrix0.access (rowIndex, termIndex) * matrix1.access (termIndex, columnIndex);
            }
        }
    }
    return result;
}

template <class U>
Matrix <U> multiply (U scalar, Matrix <U> &matrix) {
    auto result = Matrix <U> (matrix);
    for (auto entryIndex = 0; entryIndex < result.entries.size (); entryIndex++) {
        result.entries [entryIndex] *= scalar;
    }
    return result;
}

template <class U>
Matrix <U> multiply (Matrix <U> &matrix, U scalar) {
    return multiply (scalar, matrix);
}

File: zzz_voortgezet_programmeren_in_python/operator_overloading/matrix.py
class Matrix:
    def __init__ (self, nrOfRows = 1, nrOfColumns = 1, contents = None):
        if contents or matrix:
            if $$$
            self.contents = contents
            self.nrOfRows = len (self.contents)
            if self.nrOfRows:
                self.nrOfColumns = len (self.contents [0])
            else:
                self.nrOfColumns = 0
        else:
            self.nrOfRows = nrOfRows
            self.nrOfColumns = nrOfColumns
            self.contents = [[0 for columnIndex range (self.nrOfColumns)] for rowIndex in range (self.nrOfRows)]
            
    def __add__ (self, other):
    def __radd__ (self, other):
    
    def __mul__ (self, other):
        if type (other) == Matrix:
            if self.nrOfColumns == other.nrOfRows:
                result = Matrix (nrOfRows = self.nrOfRows, nrOfColumns = self.nrOfColumns)
                for rowIndex in range (self.nrOfRows):
                    for columnIndex in range (other.nrOfColumns):
                        result [rowIndex][columnIndex] = 0
                        for termIndex in range (self.nrOfColumns):
                            result [rowIndex][columnIndex] += self [rowIndex][termIndex] * other [termIndex][columnIndex]
                return result
            else:
                raise Exception ()
        else:
            result = Matrix (nrOfRows = self.nrOfRows, nrOfColumns = self.nrOfColumns
            for rowIndex in range (self.nrOfRows):
                for columnIndex in range (self.nrOfColumns):
                    contents = 
            
    
    def __rmul__ (self, other):
    
    def __pow__ (self, exponent):